Add support for database server on Ubuntu.

lwesterhof opened the pull request:
First draft for database server on Ubuntu and PgBouncer.

Vagrant file: vagrant/environments/ubuntu/Vagrantfile
Ansible: ansible-playbook playbook.yml -i environments/development/ubuntu/

- Do we want to use firewalld or ufw on Ubuntu?

dworkin committed:
YDA-4443 Experimental support for PostgreSQL 11.

dworkin committed:
YDA-4443 Silence ansible-lint.

dworkin committed:
YDA-4443 Configure ODBC for PostgreSQL 11.

dworkin committed:
YDA-4443 Make installed PostgreSQL version configurable.

dworkin committed:
YDA-4443 Silence ansible-lint.

dworkin committed:
YDA-4443 Relove one more leftover file from PostgreSQL 9.

dworkin committed:
YDA-4443 Add a final step before the database upgrade.

Which can fail.

lwesterhof committed:
Merge branch 'development' into development-postgres11

lwesterhof committed:
Merge branch 'development' into development-postgres11

lwesterhof committed:
Add support for database server on Ubuntu.

lwesterhof committed:
Use FQCN for builtin actions.

lwesterhof committed:
Fix Ansible lint issues.

lwesterhof committed:
Add missing template file.

lwesterhof committed:
Fix check mode for database roles.

lwesterhof committed:
Merge branch 'development' into development-postgres11-ubuntu

lwesterhof committed:
Fix Ansible lint issues.

lwesterhof committed:
Add role to install PgBouncer.

lwesterhof committed:
Merge branch 'development' into development-postgres11-ubuntu

lwesterhof committed:
Fix whitespace.

lwesterhof committed:
Add task to configure database host and port for iCAT.

lwesterhof committed:
Add S3 resource for Ubuntu database environment.

lwesterhof committed:
Add Ansible module to change iCAT postgres configuration.

lwesterhof committed:
Increase the maximum file descriptor number that can be opened by Minio.

lwesterhof committed:
Add TLS configuration for PgBouncer.

lwesterhof committed:
Add PgBouncer flag to iRODS iCAT role.

lwesterhof committed:
Update OIDC configuration for Ubuntu database envrionment.

lwesterhof committed:
Update PgBouncer configuration for Debian family.

lwesterhof committed:
Add PgBouncer configuration to configuring Yoda docs.

lwesterhof committed:
Merge branch 'development' into development-postgres11-ubuntu

lwesterhof committed:
Merge branch 'development' into development-postgres11-ubuntu

lwesterhof committed:
Add support for research group retention period to Yoda add group module.

lwesterhof committed:
Merge branch 'development' into development-postgres11-ubuntu

stsnel commented on environments/development/ubuntu/group_vars/ubuntu.yml:
```diff
@@ -0,0 +1,238 @@
+
+# OpenSearch module configuration
+enable_open_search: true                           # Enable OpenSearch module
+opensearch_server: portal.yoda.test
```
The opensearch_server variable needs to be changed to 'combined.yoda.test' to be able to create the index (since we don't have a separate portal server in this dev environment)

stsnel commented on environments/development/ubuntu/group_vars/ubuntu.yml:
```diff
@@ -0,0 +1,238 @@
+yoda_davrods_anonymous_enabled: true
+yoda_davrods_anonymous_fqdn: public.data.yoda.test # Yoda Davrods anonymous WebDAV fully qualified domain name (FQDN)
+
+default_yoda_schema: default-1                     # Default Yoda metadata scheme: core-0 or default-1
```
The other dev environments use the default-2 schema. Perhaps better to standardize on that.

stsnel commented on roles/postgresql/defaults/main.yml:
```diff
@@ -1,11 +1,25 @@
 
+# version of PostgreSQL to install (tested with 11-14)
+pgsql_version: 11
+prevent_upgrade: false
```
This seems like an unsafe default (at least for acc and production environments). From an admin perspective, would prefer to only initiate a database upgrade if that is explicitly configured.

stsnel commented on roles/postgresql/tasks/upgrade.yml:
```diff
@@ -0,0 +1,111 @@
+  when: "'postgresql-server' in ansible_facts.packages"
+
+
+- name: Last step before the upgrade
```
Moving this failsafe up to the top of the upgrade tasks would prevent it from leaving postgres in a stopped state. Additionally, I think it would be safer to always fail if prevent_upgrade is enabled (either "when: prevent_upgrade" or "when prevent_upgrade or ...")

stsnel commented on roles/postgresql/tasks/upgrade.yml:
```diff
@@ -0,0 +1,111 @@
+    path: "{{ item }}"
+    state: absent
+  with_items:
+    - /var/lib/pgsql/data
```
I would prefer to have a manual step in the upgrade process for removing old database data, rather than having the playbook remove it immediately. This would enable admins to verify that the upgrade has succeeded first. This way there are more options to retry / troubleshoot / revert in case something went wrong unexpectedly.

lwesterhof committed:
Only initiate a database upgrade if that is explicitly configured.

lwesterhof committed:
Update defaults for Ubuntu environment.

lwesterhof committed:
Use correct canocical module name for firewalld.

lwesterhof committed:
Moving failsafe up to the top of the upgrade tasks.

lwesterhof committed:
Add option (default false) to the upgrade process for removing old database data.

stsnel commented on roles/pgbouncer/templates/pgbouncer.ini.j2:
```diff
@@ -0,0 +1,361 @@
+;;;
+
+;; disable, allow, require, verify-ca, verify-full
+client_tls_sslmode = require
```
It looks like having client_tls_sslmode in require mode necessitates changes to the iRODS setup scripts, in order to make it work for fresh installs. See https://github.com/irods/irods/issues/6387 . So we would have to either incorporate the changes mentioned in that issue, or set client_tls_sslmode to allow.

stsnel commented on roles/pgbouncer/templates/pgbouncer.ini.j2:
```diff
@@ -0,0 +1,361 @@
+;syslog_facility = daemon
+;syslog_ident = pgbouncer
+
+;; log if client connects or server connection is made
```
Logging all connects and disconnects results in quite high log traffic. I propose turning this off, at least by default.

stsnel commented on roles/pgbouncer/templates/pgbouncer.ini.j2:
```diff
@@ -0,0 +1,361 @@
+
+;; Comma-separated list of parameters to ignore when given in startup
+;; packet.  Newer JDBC versions require the extra_float_digits here.
+;ignore_startup_parameters = extra_float_digits
```
The Ubuntu development environment needs
```
ignore_startup_parameters = extra_float_digits, geqo
```

stsnel committed:
pgbouncer: explicitly configure iCAT DB host

This change is needed to make the configuration work on
setups with a separate database server (e.g. the Ubuntu dev
environment).

stsnel committed:
pgbouncer: switch to direct MD5 authentication

The previous auth_user config required to set a password on the
Postgres account for authentication. Since iRODS uses only a single
database account, we can simplify this and just add manual
authentication for the iCAT database account instead. This way,
enabling or disabling PGBouncer does not affect the PostgreSQL
configuration.

stsnel committed:
postgresql: move package_facts task

It needs to be in main.yml instead of upgrade.yml, because
we might also need it if upgrade.yml is not included.

stsnel reviewed (commented):
It looks good to me in general!

I have added some specific feedback for a subset of pgbouncer settings, and added a couple of small proposed changes.

lwesterhof committed:
Set client_tls_sslmode to allow (workaround irods/irods#6387).

lwesterhof committed:
Do not log all PgBouncer connects and disconnects.

lwesterhof committed:
Set ignore_startup_parameters for Ubuntu environment.

stsnel reviewed (approved):
LGTM!
